File: leetcode/p62.py
# 62. Unique Paths
# https://leetcode.com/problems/unique-paths/


class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        if m==0 or n==0: return 0
        tab=[[0]*n for i in range (m)]
        for i in range(m): tab[i][0]=1
        for j in range(n): tab[0][j]=1
        for i in range(1,m):
            for j in range(1,n):
                tab[i][j]=tab[i][j-1]+tab[i-1][j]
        return tab[-1][-1]

File: leetcode/p62.cpp
class Solution {
public:
    vector<vector<int>> table;
    int uniquePaths(int m, int n) {
        table.resize(m, vector<int>(n, 0));
        table[0][0] = 1;
        for (int i=1;i<n;i++) {
            table[0][i] = table[0][i-1];
        }
        for (int i=1;i<m;i++) {
            table[i][0] = table[i-1][0];
        }
        for (int i=1;i<m;i++) {
            for (int j=1;j<n;j++) {
                table[i][j] = table[i-1][j] + table[i][j-1];
            }
        }
        return table[m-1][n-1];
    }
};
